File: LeetCode/516. Longest Palindromic Subsequence.kt
import kotlin.math.max

class Solution {
fun longestPalindromeSubseq(s:String) : Int {
        var memo = mutableMapOf<String,Int>()
        fun dp(index1:Int, index2:Int) : Int {
            if (index1 == index2) {
                return 1
            }
            else if (index1 > index2) {
                return 0
            }
            val key = index1.toString() + "|" + index2.toString()
            if (memo.containsKey(key)) {
                return memo.get(key)!!
            }
            if (s.get(index1) == s.get(index2)) {
                val result = dp(index1+1, index2-1)?.plus(2)
                return result
            }
            val result = max(dp(index1+1,index2),dp(index1,index2-1))
            memo.put(key, result)
            return result
        }
        return dp(0, s.length-1)
    }
}

File: LeetCode/516. Longest Palindromic Subsequence.py
class Solution:
    def longestPalindromeSubseq(self, s: str) -> int:
        def dp(index_1, index_2):
            if index_1 == index_2:
                return 1
            elif index_1 > index_2:
                return 0

            if memo[index_1][index_2] != -1:
                return memo[index_1][index_2]
            if s[index_1] == s[index_2]:
                memo[index_1][index_2] = dp(index_1+1, index_2-1) + 2
            else:
                memo[index_1][index_2] = max(dp(index_1+1, index_2), dp(index_1, index_2-1))
            return memo[index_1][index_2]
        memo = [[-1 for _ in range(len(s))] for _ in range(len(s))]
        result = dp(0, len(s) - 1)
        return result
